### PR TITLE
Handle corner cases for excludes

### PR DESCRIPTION
**Pull request by ljacomet**
When simplifying an AnyOf we could end with a two elements AnyOf where
one of the elements is also an AnyOf. In that case we missed the
flattening and with the right setup it can cause exclude computation to
blow up performance.
This change recognizes this corner case and special cases for it.

### CODE REVIEW

**Comment by ljacomet** [comment]
@bot-gradle test RFM

**Comment by ljacomet** [comment]
The fix was tested again a private reproducer showing the issue. A test will be added soon though it might happen after the 6.6 RC 2.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/excludes/factories/NormalizingExcludeFactory.java [hunk 1]
```diff
@@ -229,6 +230,15 @@ private ExcludeSpec doUnion(Set<ExcludeSpec> specs) {
                 elements = Arrays.stream(asArray).filter(Objects::nonNull).collect(toSet());
             }
         }
+        if (elements.size() == 2) {
+            // Corner case to handle one of the two elements being an anyOf
+            Iterator<ExcludeSpec> specIterator = elements.iterator();
+            ExcludeSpec first = specIterator.next();
+            ExcludeSpec second = specIterator.next();
+            if (first instanceof ExcludeAnyOf || second instanceof ExcludeAnyOf) {
+                return anyOf(first, second);
```
Comment (melix) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/excludes/factories/NormalizingExcludeFactory.java:239]: I wonder if we need a safe exit condition if the collection is the same as the input or we might loop.

Comment (ljacomet) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/excludes/factories/NormalizingExcludeFactory.java:239]: Damn, you are right. Went too fast on that one.
I should not delegate, I should flatten here and there.

**Comment by ljacomet** [comment]
@bot-gradle test RFM

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/excludes/factories/NormalizingExcludeFactory.java (hunk 1 above)
Comment (ljacomet) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/excludes/factories/NormalizingExcludeFactory.java]: Fixed

**Pull request merged**